walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,67 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+
+    Collections.sort(eventList, new Comparator<Event>() {
+      @Override
+      public int compare(Event event1, Event event2) {
+        return event1.getWhen().start() - event2.getWhen().start();
+      }
+    });
+
+    int currentRangeStart = TimeRange.START_OF_DAY;
+    int currentRangeEnd;
+    int currentRangeDuration;
+
+    boolean sharesAttendees;
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    for (Event event : eventList) {
+      sharesAttendees = false;
+
+      for (String person : event.getAttendees()) {
+        if (request.getAttendees().contains(person)) {
+          sharesAttendees = true;
+          break;
+        }
+      }
+
+      if (!sharesAttendees) continue;
+
+      currentRangeEnd = event.getWhen().start();
+      currentRangeDuration = currentRangeEnd - currentRangeStart;
+
+      if (currentRangeEnd >= currentRangeStart) {
+        if (currentRangeDuration >= request.getDuration()) {
+          availableTimes.add(
+            TimeRange.fromStartDuration(currentRangeStart, currentRangeDuration));
+        }
+        currentRangeStart = event.getWhen().end();
+      } else {
```

Comment:
When is this possible? Can you comment on this? I think it's when events overlap, but the names of your variables make it sound like this is impossible. Perhaps consider different names for these variables to avoid this confusion.

Reply:
You're right, I was thinking of the available time as a range with a beginning and end, but it is a bit more confusing and not very descriptive to think about it this way.